src/include/miopen/solver/implicitgemm_ck_util.hpp

```diff
@@ -157,5 +160,406 @@ ConvSolution InitAnyInvokerFactory(const ProblemDescriptionType& problem,
     return result;
 }
 
+struct TransposeInstance
+{
+    size_t tensor_sz = 0;
+    std::vector<OpKernelArg> kern_args{};
+    int index         = -1;
+    size_t buf_offset = 0;
+    shared<Data_t> buf_handle{};
+
+    template <typename TransSolnType>
+    TransposeInstance(const TransSolnType& trans_sol, int idx, const MultiBufferWorkspaceTraits& wt)
+    {
+        tensor_sz = trans_sol.GetOutputTensorSize();
+        kern_args = trans_sol.GetKernelArg();
+        index     = idx;
+        assert(index >= 0);
+        buf_offset = wt.GetOffset(index);
+    }
+
+    void AllocBuffer(const Handle& handle, Data_t workSpace)
+    {
+        buf_handle = handle.CreateSubBuffer(workSpace, buf_offset, tensor_sz);
+        assert(buf_handle.get());
+    }
+
+    Data_t GetBufferPtr() const { return buf_handle.get(); }
+
+    void ConvertFrom(const Handle& handle, const std::vector<Kernel>& kernels, ConstData_t in_ptr)
+    {
+        Run(handle, kernels, buf_handle.get(), in_ptr);
+    }
+
+    void ConvertTo(const Handle& handle, const std::vector<Kernel>& kernels, Data_t out_ptr)
+    {
+        Run(handle, kernels, out_ptr, buf_handle.get());
+    }
+
+    TransposeInstance()                         = delete;
+    TransposeInstance(const TransposeInstance&) = default;
+    TransposeInstance(TransposeInstance&&)      = default;
+    ~TransposeInstance()                        = default;
+
+private:
+    void Run(const Handle& handle,
+             const std::vector<Kernel>& kernels,
+             Data_t out_ptr,
+             ConstData_t in_ptr)
+    {
+        assert(out_ptr);
+        assert(in_ptr);
+        assert(kernels.size() > index);
+
+        kern_args[0] = out_ptr;
+        kern_args[1] = in_ptr;
+
+        handle.Run(kernels[index])(kern_args);
+        if(handle.IsProfilingEnabled())
+        {
+            handle.AccumKernelTime(handle.GetKernelTime());
+        }
```

Comment:
IIRC This won't work because the legacy internal API for accumulation of elapsed kernel times is cumbersome. Running the kernel overwrites the value stored in the handle by `AccumKernelTime()`, i.e. when you run the 2nd kernel, the elapsed time of the 1st kernel is lost. T. Example of valid use:
```C++
float accumulated = 0;
... // Run the 1st kernel, then:
if(handle.IsProfilingEnabled())
    accumulated += handle.GetKernelTime();
...
... // Run the Nth kernel, then:
if(handle.IsProfilingEnabled())
    accumulated += handle.GetKernelTime();
...
if(handle.IsProfilingEnabled())
{
    handle.ResetKernelTime();
    handle.AccumKernelTime(accumulated);
}
```
So the recommendation is: move accumulation of elapsed kernel times up, into the Invoker.

Or you can `ResetKernelTime()` somewhere in the Invoker, before running the 1st kernel, and do here like this:
```C++
auto save = handle.IsProfilingEnabled() ? 0.0f : handle.GetKernelTime(); // save previously accumulated time
handle.Run(kernels[index])(kern_args);
if(handle.IsProfilingEnabled())
    handle.AccumKernelTime(save); // add prev accumulated time to the execution time of kernels[index].
```
This approach allows using the value stored in the handle as an accumulator.